cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataEntity.java

```diff
@@ -89,13 +100,117 @@ public MetadataEntity append(String key, String value) {
     return new MetadataEntity(existingParts);
   }
 
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    MetadataEntity that = (MetadataEntity) o;
+    return Objects.equals(details, that.details);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(details);
+  }
+
+  @Override
+  public String toString() {
+    return "MetadataEntity{" +
+      "details=" + details +
+      '}';
+  }
+
   /**
    * @return A {@link List} of {@link KeyValue} representing the metadata entity
    */
   public List<KeyValue> getKeyValues() {
     return details;
   }
 
+  /**
+   * @return A {@link Iterator} over the keys in this {@link MetadataEntity}
+   */
+  public Iterator<String> getKeys() {
```

Comment:
Why return an `Iterator` instead of `Iterable`? You have everything in memory already, returning an `Iterable` is easier to consume.